[receiver/haproxyreceiver] add more metrics

atoulme opened the pull request:
**Description:**
Add a lot more metrics to the haproxyreceiver. Use attributes as resource attributes if possible.

**Link to tracking Issue:**
#16829

**Testing:**
Unit tests.

**Documentation:**
Generated from yaml.

atoulme committed:
[receiver/haproxyreceiver] add more metrics

MovieStoreGuy commented on receiver/haproxyreceiver/metadata.yaml:
```diff
@@ -22,49 +22,241 @@ resource_attributes:
 
+attributes:
+  status_code:
+    description: Status code category, 1xx, 2xx, 3xx, 4xx, 5xx or other
```
You can set this to be an enum value from memory. 

I think the structure is:

```yaml
attributes:
   status_code:
      description: Status code group
      type: string
      enum:
      - "1xx"
      - "2xx"
      - "3xx"
      - "4xx"
      - "5xx"
```

MovieStoreGuy commented on receiver/haproxyreceiver/metadata.yaml:
```diff
@@ -22,49 +22,241 @@ resource_attributes:
+
 metrics:
-  haproxy.connection_rate:
+  haproxy.connections.rate:
```
I appreciate the move to `.` spacing, thank you.

MovieStoreGuy commented on receiver/haproxyreceiver/metadata.yaml:
```diff
@@ -22,49 +22,241 @@ resource_attributes:
+      input_type: string
+    unit: "{sessions}"
+  haproxy.connections.total:
+    description: Cumulative number of connections (frontend). Corresponds to HAProxy's `conn_tot` metric.
```
Is it worth referencing the HA Proxy's documentation for those who want to read further about the original / underlying metric using the `extended_documentation` field?

atoulme replied:
I can't find an authoritative documentation for haproxy metrics :(

MovieStoreGuy reviewed (commented):
Loving the increased addition of metrics, however, instead of using the description to highlight the relation to the actual HAProxy metric, lets use the `extended_documentation` field to link back to the HA Proxy docs so that any and all possible questions could be explained about what the original value means / comes from.

atoulme committed:
add more checks and use enum

atoulme commented:
Thanks for the early review! Much appreciated. I was pushing last night and updated dependencies when I saw there was a conflict - it looks like the new dependencies introduced this new explicit error for empty value. I have fixed those now. 

Agree on linking to haproxy, but not sure on a canonical place. I will have to look.

atoulme commented:
I have not found a source of truth for haproxy metrics. Can we continue the review and revisit later?

bogdandrutu commented:
Needs rebase
